

#' Generate a cnv Granges from a tab delimited created by ASCAT
#' @param ascat.fn input dataframe with CNV segments generated by ASCAT,
#' @return A GRanges object with the segments predicted by ASCAT
#' @export
parseASCAT_file=function(ascat.fn){
  ascat_tn=read.csv(file=ascat.fn,
                    header=FALSE)
  colnames(ascat_tn)=c(
    "Segment number",
    "Chromosome",
    "Start",
    "End",
    "Total copy number—normal",
    "Minor copy number—normal",
    "Total copy number—tumour",
    "Minor copy number—tumour"
  )
  return( parseASCAT(ascat_tn))
}

#' Generate a cnv Granges from a dataframe created by ASCAT
#' @param ascat.df input dataframe with CNV segments generated by ASCAT,
#' @return A GRanges object with the segments predicted by ASCAT
#' @export
parseASCAT=function(ascat.df){

  drop=which(ascat.df$`Total copy number—tumour`==2 & ascat.df$`Minor copy number—tumour`==1)
  if(length(drop)>0){
    ascat.df=ascat.df[ -drop , ]
  }

  ascat.gr=GenomicRanges::makeGRangesFromDataFrame(ascat.df,
                                                   keep.extra.columns = TRUE,
                                                   ignore.strand=TRUE)%>%
    GenomeInfoDb::sortSeqlevels( ) %>%
    BiocGenerics::sort()

  GenomeInfoDb::seqlevelsStyle(ascat.gr)='UCSC'
  #get the info from the file
  mcols_new=S4Vectors::mcols( ascat.gr ) %>%
    tibble::as_tibble() %>%
    dplyr::mutate( total.allele.copies=`Total copy number—tumour`,
                   minor.allele.copies=`Minor copy number—tumour`,
                   major.allele.copies=total.allele.copies-minor.allele.copies,)

  flags=apply(mcols_new, 1, function(K){ cnvFlag(total.cn = K['total.allele.copies'], minor.cn = K['minor.allele.copies'] )}) # pass the total.cn to make sure that cases with uncertainty=-1 are parsed properly
  mcols_new = mcols_new %>%
    dplyr::mutate( cnv.flag=flags)
  S4Vectors::mcols( ascat.gr)=mcols_new

  return(ascat.gr)
}

